pkg/asset/storage.go

```diff
@@ -0,0 +1,369 @@
+		return fmt.Errorf("failed to query object list: %w", err)
+	}
+	toDelete := []string{}
+	intersection := 0
+	for _, obj := range existing {
+		keep := false
+		if time.Since(obj.createdAt) < deletionEmbargo {
+			// To avoid races between object upload and object deletion, we don't delete
+			// newly uploaded files for a while after they're uploaded.
+			keep = true
+		}
+		if val, ok := needed[obj.downloadURL]; ok && val {
+			keep = true
+			intersection++
+		}
+		storage.tracer.Log("-- object %v, %v: keep %t", obj.downloadURL, obj.createdAt, keep)
+		if !keep {
+			toDelete = append(toDelete, obj.downloadURL)
+		}
+	}
+	const intersectionCheckCutOff = 4
+	if len(existing) > intersectionCheckCutOff && intersection == 0 {
+		// This is a last-resort protection against possible dashboard bugs.
+		// If the needed assets have no intersection with the existing assets,
+		// don't delete anything. Otherwise, if it was a bug, we will lose all files.
+		return fmt.Errorf("needed assets have some intersection with the existing ones")
+	}
+	for _, url := range toDelete {
+		err := storage.backend.remove(url)
+		storage.tracer.Log("-- deleted %v: %v", url, err)
+		// Several syz-ci's might be sharing the same storage. So let's tolerate
+		// races during file deletion.
+		if err != nil && err != ErrAssetDoesNotExist {
+			return fmt.Errorf("asset deletion failure: %w", err)
+		}
+	}
+	return nil
+}
+
+type uploadRequest struct {
+	origFile        string
+	savePath        string
+	contentEncoding string
+	contentType     string
+}
+
+type uploadResponse struct {
+	downloadURL string
+}
+
+type storedObject struct {
+	downloadURL string
+	createdAt   time.Time
+}
+
+type StorageBackend interface {
+	upload(req *uploadRequest) (*uploadResponse, error)
+	list() ([]storedObject, error)
+	remove(path string) error
+}
+
+type PreprocessorCollection struct {
+	defaultOne Preprocessor
+	custom     map[Type]Preprocessor
+	// Map cannot be concurrently accessed.
+	mu sync.Mutex
+}
+
+func MakePreprocessorCollection(defaultOne Preprocessor,
+	custom map[Type]Preprocessor) *PreprocessorCollection {
+	ret := &PreprocessorCollection{
+		defaultOne: PreprocessGzip,
+		custom:     custom,
+	}
+	for t, p := range custom {
+		if p.Available == nil || p.Available() {
+			ret.custom[t] = p
+		}
+	}
+	return ret
+}
+
+func (ap *PreprocessorCollection) GetPreprocessor(assetType Type) Preprocessor {
+	ap.mu.Lock()
+	defer ap.mu.Unlock()
+	p, ok := ap.custom[assetType]
+	if ok {
+		return p
+	}
+	return ap.defaultOne
+}
+
+type Preprocessor struct {
+	Available func() bool
+	Do        func(req *uploadRequest,
+		next func(req *uploadRequest) (*uploadResponse, error)) (*uploadResponse, error)
+}
+
+var xzPresent bool
+var xzCheck sync.Once
+
+const xzTimeout = 5 * time.Minute
+const xzCompressionRatio = 0
+const xzThreadsCount = 6
+
+var PreprocessXz = Preprocessor{
+	Available: func() bool {
+		xzCheck.Do(func() {
+			_, err := osutil.RunCmd(time.Minute, "", "xz --version")
```

Comment:
Have you considered using some Go package for this?
The tool needs to be installed separately, not sure if it's available on *BSD, Mac, Windows, Fuchsia. May have different interface there.
If we do this, it will also remove all of the Available-related code code.